spring-cloud-function-adapters/spring-cloud-function-adapter-gcp/src/test/java/org/springframework/cloud/function/adapter/gcloud/FunctionInvokerBackgroundTests.java

```diff
@@ -0,0 +1,296 @@
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import com.google.gson.Gson;
+import org.junit.Ignore;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.contrib.java.lang.system.SystemOutRule;
+
+import org.springframework.cloud.function.context.config.ContextFunctionCatalogAutoConfiguration;
+import org.springframework.cloud.function.json.JsonMapper;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.context.annotation.Import;
+import org.springframework.messaging.Message;
+import org.springframework.messaging.converter.AbstractMessageConverter;
+import org.springframework.messaging.converter.MessageConverter;
+import org.springframework.messaging.support.MessageBuilder;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * Unit tests for the background functions adapter for Google Cloud Functions.
+ *
+ * @author Dmitry Solomakha
+ * @author Mike Eltsufin
+ */
+public class FunctionInvokerBackgroundTests {
+
+	private static final Gson gson = new Gson();
+
+	@Rule
+	public final SystemOutRule systemOutRule = new SystemOutRule().enableLog();
+
+	private static final String DROPPED_LOG_PREFIX = "Dropping background function result: ";
+
+	@Test
+	public void testHelloWorldSupplier_Background() throws Exception {
+		testBackgroundFunction(HelloWorldSupplier.class, null, "Hello World!", null, null);
+	}
+
+	@Test
+	public void testJsonInputFunction_Background() throws Exception {
+		testBackgroundFunction(JsonInputFunction.class, new IncomingRequest("hello"),
+				"Thank you for sending the message: hello", null, null);
+	}
+
+	@Test
+	public void testJsonInputOutputFunction_Background() throws Exception {
+		testBackgroundFunction(JsonInputOutputFunction.class, new IncomingRequest("hello"),
+				new OutgoingResponse("Thank you for sending the message: hello"), null, null);
+	}
+
+	@Test
+	public void testJsonInputConsumer() throws Exception {
+		testBackgroundFunction(JsonInputConsumer.class, new IncomingRequest("hello"), null,
+				"Thank you for sending the message: hello", null);
+	}
+
+	@Test
+	public void testPubSubBackgroundFunction_PubSub() throws Exception {
+		PubSubMessage pubSubMessage = new PubSubMessage();
+		pubSubMessage.setData("hello");
+		testBackgroundFunction(PubsubBackgroundFunction.class, pubSubMessage, null,
+				"Thank you for sending the message: hello", "google.pubsub.topic.publish");
+	}
+
+	@Test
+	public void testPubSubBackgroundFunction_PubSubPayload() throws Exception {
+		PubSubMessage pubSubMessage = new PubSubMessage();
+		IncomingRequest message = new IncomingRequest("Hello");
+		pubSubMessage.setData(gson.toJson(message));
+		testBackgroundFunction(PubsubBackgroundFunctionPayload.class, pubSubMessage, null,
+				"Thank you for sending the message: Hello", "google.pubsub.topic.publish");
+	}
+
+	@Test
+	@Ignore("Conversion to String is not supported yet.")
```

Comment:
@olegz See: #503 .

Reply:
@dzou @dmitry-s Oleg suggested to look at `BeanFactoryAwareFunctionRegistryTests.testSerializationWithCompatibleWildcardSubtypeAcceptHeader()` for  an example.